apps/api/src/modules/payments/payments.service.ts

```diff
@@ -541,43 +544,75 @@ export default class PaymentsService {
     // Attempt to find card (cardId could be source or db ID)
     const card = await PaymentCardModel.query(trx).findById(cardId)
 
-    // Create payment using Circle API
-    const payment = await this.circle
+    // Circle only accepts loopback addresses
+    const verificationHostname = Configuration.webUrl.includes('localhost')
+      ? 'http://127.0.0.1:3000'
+      : Configuration.webUrl
+
+    // Base payment details
+    const basePayment = {
+      idempotencyKey,
+      metadata: {
+        ...metadata,
+        sessionId: SHA256(user.id).toString(enc.Base64),
+      },
+      amount: {
+        amount: priceInUSD,
+        currency: DEFAULT_CURRENCY,
+      },
+      description,
+      source: {
+        id: card?.externalId || cardId,
+        type: CirclePaymentSourceType.card, // @TODO: Update when support ACH
+      },
+    }
+
+    // Create 3DS payment
+    const paymentResponse = await this.circle
       .createPayment({
-        idempotencyKey: idempotencyKey,
-        metadata: metadata,
-        amount: {
-          amount: priceInUSD,
-          currency: DEFAULT_CURRENCY,
-        },
-        verification: verification,
-        description: description,
-        source: {
-          id: card?.externalId || cardId,
-          type: CirclePaymentSourceType.card, // @TODO: Update when support ACH
-        },
+        ...basePayment,
         ...encryptedDetails,
+        verification: CirclePaymentVerificationOptions.three_d_secure,
+        verificationSuccessUrl: new URL(
+          Configuration.successPath,
+          verificationHostname
+        ).toString(),
+        verificationFailureUrl: new URL(
+          Configuration.failurePath,
+          verificationHostname
+        ).toString(),
       })
-      .catch((error) => {
-        this.logger.error(error, 'failed to create payment')
-        return null
-      })
-
-    if (!payment) {
+      .catch(() => null)
+
+    let payment: ToPaymentBase | undefined
+    if (!paymentResponse) {
+      // Create cvv payment
+      const cvvPayment = await this.circle
+        .createPayment({
+          ...basePayment,
+          ...encryptedDetails,
+          verification: CirclePaymentVerificationOptions.cvv,
+        })
+        .catch(() => null)
```

Comment:
Need to verify Circle will immediately fail, and not fail on the verification failure URL page, so that we can run cvv verification after. Through testing I was only finding that errors resulted in the failure URL being hit, but talking with Circle initially, we expected to get an immediate failure so we can run secondary verification, so waiting to hear back. 🤞